pkg/scheduler/algorithm/priorities/scarce_resource_bin_packing.go

```diff
@@ -0,0 +1,126 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package priorities
+
+import (
+	"fmt"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/kubernetes/pkg/scheduler/algorithm"
+	schedulerapi "k8s.io/kubernetes/pkg/scheduler/api"
+	schedulercache "k8s.io/kubernetes/pkg/scheduler/cache"
+)
+
+// ScarceResourceBinPacking contains information to calculate bin packing priority.
+type ScarceResourceBinPacking struct {
+	scarceResource string
+}
+
+// NewScarceResourceBinPacking creates a ScarceResourceBinPackingPriorityMap.
+func NewScarceResourceBinPacking(scarceResource string) (algorithm.PriorityMapFunction, algorithm.PriorityReduceFunction) {
+	scarceResourceBinPackingPrioritizer := &ScarceResourceBinPacking{
+		scarceResource: scarceResource,
+	}
+	return scarceResourceBinPackingPrioritizer.ScarceResourceBinPackingPriorityMap, nil
+}
+
+// ScarceResourceBinPackingPriorityDefault creates a requestedToCapacity based
+// ResourceAllocationPriority using default resource scoring function shape.
+// The default function assigns 1.0 to resource when all capacity is available
+// and 0.0 when requested amount is equal to capacity.
+func ScarceResourceBinPackingPriorityDefault() *ScarceResourceBinPacking {
+	defaultscarceResourceBinPackingPrioritizer := &ScarceResourceBinPacking{
+		scarceResource: "",
+	}
+	return defaultscarceResourceBinPackingPrioritizer
+}
+
+// ScarceResourceBinPackingPriorityMap is a priority function that favors nodes that have higher utlization of scare resource.
+// It will detect whether the requested scarce resource is present on a node, and then calculate a score ranging from 0 to 10
+// based total utlization (best fit)
+// - If none of the scare resource are requested, this node will be given the lowest priority.
+// - If the scarce resource is requested, the larger the resource utlization ratio, the higher the node's priority.
+func (s *ScarceResourceBinPacking) ScarceResourceBinPackingPriorityMap(pod *v1.Pod, meta interface{}, nodeInfo *schedulercache.NodeInfo) (schedulerapi.HostPriority, error) {
+	var score int
+	node := nodeInfo.Node()
+	if len(s.scarceResource) == 0 {
+		return schedulerapi.HostPriority{}, fmt.Errorf("scarceResource not defined")
+	}
+	if node == nil {
+		return schedulerapi.HostPriority{}, fmt.Errorf("node not found")
+	}
+	if !podRequestsResource(*pod, s.scarceResource) {
+		score = 0
+	} else {
+		score = calculateScareResourceScore(nodeInfo, pod.Spec.Containers, s.scarceResource)
```

Comment:
Don't forget to take the max of this value and the calculated score over the set of "init containers". See `podRequestsResource` helper function below, which does look at init containers.

Reply:
Should we consider init containers in prioritizes? "init containers" is a "short running" work to me, I think we'd like to get better resource utilization for main container.

We need to consider "init containers" in predicates to avoid resource exhausted.